components/core/src/package/ident.rs

```diff
@@ -304,12 +299,40 @@ impl<'a> From<&'a PackageIdent> for Cow<'a, PackageIdent> {
 pub struct FullyQualifiedPackageIdent(PackageIdent);
 
 impl FullyQualifiedPackageIdent {
+    /// Creates a new fully qualified package identifier
+    pub fn new<T: Into<String>>(origin: T, name: T, version: T, release: T) -> Self {
+        let ident = PackageIdent { origin:  origin.into(),
+                                   name:    name.into(),
+                                   version: Some(version.into()),
+                                   release: Some(release.into()), };
+        FullyQualifiedPackageIdent(ident)
+    }
+
     pub fn archive_name(&self) -> String {
         self.0
-            .as_ref()
             .archive_name()
             .unwrap_or_else(|_| panic!("PackageIdent {} should be fully qualified", self.0))
     }
+
+    pub fn guaranteed_version(&self) -> &str {
+        self.version()
+            .unwrap_or_else(|| panic!("PackageIdent {} should be fully qualified", self.0))
+    }
+
+    pub fn guaranteed_release(&self) -> &str {
+        self.release()
+            .unwrap_or_else(|| panic!("PackageIdent {} should be fully qualified", self.0))
+    }
+}
```

Comment:
This API seems a little "off" to me... I would expect that these would be guaranteed by virtue of being methods on a `FullyQualifiedPackageIdent`, so I would expect just `version()` and `release()`. Is there a reason that I'm missing that makes the `guaranteed_` prefix more useful?

I assume that the fact that we keep a `PackageIdent` inside the struct is to help inherit the various comparison trait implementations, but I can imagine a future where we just keep the individual components as fields within this struct, which would allow us to get rid of the `unwrap_or_else` here. I'm cool with keeping this for now, though, since this is transitional, and there's a lot more work to do.

Reply:
Possibly dumb related question: Are the guaranteed_*  methods needed because we implement Identifiable, which returns Option<string> for the release and version methods? If we implemented release() -> string that would have a different type sig, and I don't fully understand rust's type rules.

Reply:
Strictly speaking, no, but I think issues arise from using `Identifiable#version` in this implementation.

(This could also be a suggestion we postpone until later, since there will be more work on this.)

Reply:
> This API seems a little "off" to me

Agreed. Addressed in [ad924b](https://github.com/habitat-sh/habitat/pull/7507/commits/ad924bf0e2062c5f15a01877d3ff4de22abcf13e). This was me just trying to get something to compile and then forgetting to come back and clean it up. Thanks!

> I assume that the fact that we keep a PackageIdent inside the struct is to help inherit the various comparison trait implementations, but I can imagine a future where we just keep the individual components as fields within this struct,

Right... Agreed, that sounds like a much better future.